File: Java/Odometer.java
package Odometer;

import java.util.ArrayList;

public class Odometer {
    private int size = 0;
    private int length = 0;
    private int position = 0;
    private ArrayList<Integer> readings = new ArrayList<Integer>();

    private boolean isAscending(int n) {
        if (n < 10) 
            return true;
        if ((n % 10) <= (n / 10) % 10) 
            return false;
        return isAscending(n / 10);
    }

    public Odometer(int size) {
        if (size < 2 || size > 8) return;

	    this.size = size;

	    int start = 1;
        for (int i = 1; i < size; ++i) 
            start *= 10;

	    for (int i = start; i < start * 10; ++i) {
	        if (isAscending(i)) {
		        this.readings.add(i);
		        this.length++;
	        }
	    }
    }

	int next(int step) {
        for (int i = 0; i < step; ++i) 
            this.position++;

        while (this.position >= this.length) 
            this.position -= this.length;
            
	    return this.readings.get(this.position);
	}

	int prev(int step) {
	    for (int i = 0; i < step; ++i) 
                 this.position--;

	    while (this.position < 0) 
		    this.position += this.length;

	    return this.readings.get(this.position);
	}

	int diff(Odometer b) {
	    if (this.size != b.size) return -1;

	    if (this.position >= b.position) 
		    return this.position - b.position;
	    else 
		    return this.position - b.position + this.length;
    }
}

File: Python/Odometer.py
class Odometer:
    DIGITS = '123456789'
    def __init__(self, size):
        self.START = int(Odometer.DIGITS[:size])
        self.LIMIT = int(Odometer.DIGITS[-size:])
        self.SIZE = size
        self.readings = []
        for n in range(10 ** (size - 1), 10 ** size):
            if self.is_ascending(n):
                self.readings.append(n) 

    def is_ascending(self, n : int) -> bool:
        return n == int(''.join(sorted(str(n))))


    def next_reading(self, n : int) -> int:
        if n == self.LIMIT:
            return self.START
        n += 1
        while self.is_ascending(n) == False:
            n += 1
        return n 


    def prev_reading(self, n : int) -> int:
        if n == self.START:
            return self.LIMIT
        n -= 1
        while self.is_ascending(n) == False:
            n -= 1
        return n


    def next_nthReading(self, n : int, step : int) -> int:
        for _ in range(step):
            n = self.next_reading(n)
        return n


    def prev_nthReading(self, n : int, step : int) -> int:
        for _ in range(step):
            n = self.prev_reading(n)
        return n


    def readings_in_range(self, start : int, limit : int) -> list:
        return self.readings[start : limit + 1]